nizarmah/lace
# Support single test multiple cases

### Description

* Separates framework and tests configs (ts + eslint)

* Cleans existing test before generation

* Refactors test generation
    * Uses generators heavily
    * Clones the function for each test
    * Extracts assertions into an imported module

* Supports multiple test cases

### Testing

1. Run:
    
    ```shell
    npm run test:hello:gen:test && npm run test:hello:run
    
    npm run test:greet:gen:test && npm run test:greet:run
    
    npm run test:calculator:gen:test && npm run test:calculator:run
    ```

1. Ensure no errors

1. Open: `tests/*.ts`

1. Ensure they look fine

### Example

Generated code for calculator.

```typescript
function parseTokens(expression: string): string[] {
    return expression.match(/(\d+\.?\d*|\+|-|\*|\/)/g) ?? [];
}
function __parseTokens__test__0(expression: string): string[] {
    const __lace__proxy__returns = expression.match(/(\d+\.?\d*|\+|-|\*|\/)/g) ?? [];
    __lace__assert__equal(__lace__proxy__returns, ["1", "+", "2", "-", "3", "*", "4", "/", "5"]);
    return __lace__proxy__returns;
}
__parseTokens__test__0("1 + 2 - 3 * 4 / 5");
function __parseTokens__test__1(expression: string): string[] {
    const __lace__proxy__returns = expression.match(/(\d+\.?\d*|\+|-|\*|\/)/g) ?? [];
    __lace__assert__equal(__lace__proxy__returns, []);
    return __lace__proxy__returns;
}
__parseTokens__test__1("");
function __parseTokens__test__2(expression: string): string[] {
    const __lace__proxy__returns = expression.match(/(\d+\.?\d*|\+|-|\*|\/)/g) ?? [];
    __lace__assert__equal(__lace__proxy__returns, []);
    return __lace__proxy__returns;
}
__parseTokens__test__2("hello?");
```
